File: flutter_planbook/packages/planbook_widget/android/src/main/kotlin/com/bapaws/planbook_widget/EngineReadyTracker.kt
package com.bapaws.planbook_widget

import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import kotlin.coroutines.resume

/**
 * Flutter 引擎就绪状态追踪器。
 *
 * 工作流程：
 *   1. [PlanbookWidgetPlugin.onAttachedToEngine] 时通过 [attach] 注册 channel；
 *   2. Dart 端在 handler 注册完成后调用 `notifyReady`，触发 [markReady]；
 *   3. 调用方（[WidgetActionDispatcher]）通过 [awaitChannel] 等待两件事都满足后
 *      拿到 channel 进行 invokeMethod。
 *
 * 进程内单例，FlutterEngineCache 和当前进程绑定（headless engine 也使用同一个实例）。
 */
internal object EngineReadyTracker {

    private val lock = Any()

    @Volatile
    private var channel: MethodChannel? = null

    @Volatile
    private var isReady = false

    private val waiters = mutableListOf<(MethodChannel?) -> Unit>()

    /** 由 plugin attach 时调用 */
    fun attach(c: MethodChannel) {
        synchronized(lock) {
            channel = c
        }
        flushWaitersIfReady()
    }

    /** 由 plugin detach 时调用 */
    fun detach() {
        val pending: List<(MethodChannel?) -> Unit>
        synchronized(lock) {
            channel = null
            isReady = false
            pending = waiters.toList()
            waiters.clear()
        }
        pending.forEach { it.invoke(null) }
    }

    /** Dart 端 `notifyReady` 触发 */
    fun markReady() {
        synchronized(lock) {
            isReady = true
        }
        flushWaitersIfReady()
    }

    private fun flushWaitersIfReady() {
        val ch: MethodChannel?
        val toResume: List<(MethodChannel?) -> Unit>
        synchronized(lock) {
            if (!isReady || channel == null) return
            ch = channel
            toResume = waiters.toList()
            waiters.clear()
        }
        toResume.forEach { it.invoke(ch) }
    }

    /**
     * 等待 channel 就绪。超时返回 null。
     * 调用方应在协程作用域内执行。
     */
    suspend fun awaitChannel(timeoutMs: Long): MethodChannel? =
        withTimeoutOrNull(timeoutMs) {
            synchronized(lock) {
                if (isReady && channel != null) return@withTimeoutOrNull channel
            }
            suspendCancellableCoroutine<MethodChannel?> { cont ->
                val callback: (MethodChannel?) -> Unit = { ch ->
                    if (cont.isActive) cont.resume(ch)
                }
                synchronized(lock) {
                    if (isReady && channel != null) {
                        cont.resume(channel)
                        return@suspendCancellableCoroutine
                    }
                    waiters.add(callback)
                }
                cont.invokeOnCancellation {
                    synchronized(lock) { waiters.remove(callback) }
                }
            }
        }
}

File: flutter_planbook/packages/planbook_widget/ios/planbook_widget/Sources/planbook_widget/EngineReadyTracker.swift
import Foundation
import Flutter

/// Flutter 引擎就绪状态追踪器（iOS）。
///
/// 工作流程：
///   1. `PlanbookWidgetPlugin.register(with:)` 时通过 `attach(channel:)` 注册 channel；
///   2. Dart 端注册完 handler 后调用 `notifyReady`，触发 `markReady()`；
///   3. 调用方（如 `WidgetActionDispatcher`）通过 `awaitChannel(timeout:)` 拿到 channel。
///
/// 单例线程安全；channel 与 ready 信号都满足后才会唤醒等待者。
public final class EngineReadyTracker {

    public static let shared = EngineReadyTracker()

    private let lock = NSLock()
    private var channel: FlutterMethodChannel?
    private var isReady: Bool = false
    private var waiters: [(FlutterMethodChannel?) -> Void] = []

    private init() {}

    /// Plugin attach 时调用。
    public func attach(channel: FlutterMethodChannel) {
        lock.lock()
        self.channel = channel
        lock.unlock()
        flushWaitersIfReady()
    }

    /// Plugin detach 时调用，清空状态以避免下一次进程冷启动时拿到陈旧引用。
    public func detach() {
        lock.lock()
        channel = nil
        isReady = false
        let pending = waiters
        waiters.removeAll()
        lock.unlock()
        pending.forEach { $0(nil) }
    }

    /// Dart 端 `notifyReady` 触发。
    public func markReady() {
        lock.lock()
        isReady = true
        lock.unlock()
        flushWaitersIfReady()
    }

    private func flushWaitersIfReady() {
        lock.lock()
        guard isReady, let ch = channel else {
            lock.unlock()
            return
        }
        let toResume = waiters
        waiters.removeAll()
        lock.unlock()
        toResume.forEach { $0(ch) }
    }

    /// 等待 channel 就绪。超时返回 nil。
    public func awaitChannel(timeout: TimeInterval) async -> FlutterMethodChannel? {
        // 快路径
        lock.lock()
        if isReady, let ch = channel {
            lock.unlock()
            return ch
        }
        lock.unlock()

        return await withTaskGroup(of: FlutterMethodChannel?.self) { group in
            group.addTask { [weak self] in
                await withCheckedContinuation { (cont: CheckedContinuation<FlutterMethodChannel?, Never>) in
                    guard let self = self else {
                        cont.resume(returning: nil)
                        return
                    }
                    self.lock.lock()
                    if self.isReady, let ch = self.channel {
                        self.lock.unlock()
                        cont.resume(returning: ch)
                        return
                    }
                    self.waiters.append { ch in cont.resume(returning: ch) }
                    self.lock.unlock()
                }
            }
            group.addTask {
                try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                return nil
            }
            let result = await group.next() ?? nil
            group.cancelAll()
            return result
        }
    }
}
